### ISSUE TITLE
Translations

### ISSUE THREAD

**Issue by jicho** [simojenki/bonob#13]
Hi,

I just did another "dive" into bonob.
When looking at the "catalog" under bonob under the music icon in the Sonos app I have English and Dutch items mixed.
Is it possible to translate, for example, the "Starred" category in the list?

Or... is this linked to the Navidrome language settings? In case it is I don't understand the mix of English and Dutch.

Just asking so life can be easier for family members when I'm doing playing/testing.

Thanks for the reaction.

**Comment by simojenki** [comment]
I'm not sure I understand how you have Dutch and English, could you share a screen shot somehow?

Generally speaking it is possible to translate the menu options.  ie. Artists, Albums, ..., Starred.  This is done using the sonos configuration, the language is taken from your sonos app, which must come from you device somehow.  

I can take a look at how that works if you are able to provide the appropriate dutch translations?

**Comment by jicho** [comment]
Hi,

Here is a screenshot:

![Schermafbeelding 2021-08-03 om 07 43 22](https://user-images.githubusercontent.com/610740/127966815-03ec5fcf-df84-434a-91c1-5a1057f90a02.png)

I must admit, I mixed it up... It is probably all English... Some words are just the same in Dutch...

On the other hand, a translation would be appreciated.
This is a mapping, mostly, based on the Navidrome translations:

| *EN*            | *NL*               |
|-----------------|--------------------|
| Artists         | Artiesten          |
| Albums          | Albums             |
| Playlists       | Afspeellijsten     |
| Genres          | Genres             |
| Random          | Willekeurig        |
| Starred         | Favorieten         |
| Recently Added  | Onlangs toegevoegd |
| Recently Played | Onlangs afgespeeld |
| Most Played     | Meest afgespeeld   |

Hope you can add it.

**Comment by simojenki** [comment]
This is mostly done.

Perhaps you could translate some more

**Comment by simojenki** [comment]
What about "Tracks"?

**Comment by simojenki** [comment]
and more..
- Success
- Failure
- Expected configuration
- Existing service config
- No existing service registration
- Register
- Remove Registration
- Devices
- Services
- Login
- Log in to bonob
- Username
- Password
- Successfully registered
- Registration failed!
- Successfully removed registration
- Failed to remove registration!
- Invalid linkCode!
- Login successful!
- Login failed!
- Linking sonos with bonob

**Comment by jicho** [comment]
Nice!

Based on what I've seen on the screens in the Sonos App, the bonob configuration page and during connection I think this are (for now) the best translations:

| *EN*                              | *NL*                              |
|-----------------------------------|-----------------------------------|
| Tracks                            | Nummers                           |
| Success                           | Gelukt                            |
| Failure                           | Mislukt                           |
| Expected configuration            | Verwachte configuratie            |
| Existing service config           | Bestaande serviceconfiguratie     |
| No existing service registration  | Geen bestaande serviceregistratie |
| Register                          | Registreren                       |
| Remove Registration               | Verwijder registratie             |
| Devices                           | Apparaten                         |
| Services                          | Services                          |
| Login                             | Inloggen                          |
| Log in to bonob                   | Login op bonob                    |
| Username                          | Gebruikersnaam                    |
| Password                          | Wachtwoord                        |
| Successfully registered           | Registratie geslaagd              |
| Registration failed!              | Registratie mislukt!              |
| Successfully removed registration | Registratie succesvol verwijderd  |
| Failed to remove registration!    | Kon registratie niet verwijderen! |
| Invalid linkCode!                 | Ongeldige linkcode!               |
| Login successful!                 | Inloggen gelukt!                  |
| Login failed!                     | Inloggen mislukt!                 |
| Linking sonos with bonob          | Koppel sonos met bonob            |

Oh you might think of changing this ```Remove Registration``` into ```Remove registration```. Two words that start both with a capital is more a thing for programmers ;)

Good luck and let me know when you need more translations

**Comment by simojenki** [comment]
I've applied nl-NL translations, pls try again

**Comment by jicho** [comment]
On the login page everything is still in English, same in the app (tried on two iOS devices).

Is there anything that I can check for you to see if bonob is checking to get the NL translation?

**Comment by jicho** [comment]
Just found the strings XML!

When I connect the service I see the translation, this one in my case:
```
<string stringId="AppLinkMessage">
<![CDATA[ Sonos koppelen aan musicbox ]]>
</string>
```

So that seems to work...
For some reason the overview (see my screenshot above) is still in EN.
Can this be caching?

**Comment by jicho** [comment]
Maybe this helps...

After this request:
```
{
  "message": "Fetching metadata type=root, typeId=undefined, lang=(key) => {\n        var _a;\n        const value = (_a = langToUse[key]) === null || _a === void 0 ? void 0 : _a.replace(\"$BONOB_SONOS_SERVICE_NAME\", serviceName);\n        if (value)\n            return value;\n        else\n            throw `No translation found for ${langs}:${key}`;\n    }",
  "level": "debug",
  "service": "bonob",
  "timestamp": "2021-08-14 12:36:38"
}
```

The following in the log is:
```
{
  "message": {
    "level": "debug",
    "data": "<?xml version=\"1.0\" encoding=\"utf-8\"?><soap:Envelope xmlns:soap=\"http://schemas.xmlsoap.org/soap/envelope/\"  xmlns:tns=\"http://www.sonos.com/Services/1.1\"><soap:Body><getMetadataResponse xmlns=\"http://www.sonos.com/Services/1.1\"><getMetadataResult><count>9</count><index>0</index><total>9</total><mediaCollection><itemType>container</itemType><id>artists</id><title>Artists</title></mediaCollection><mediaCollection><itemType>albumList</itemType><id>albums</id><title>Albums</title></mediaCollection><mediaCollection readOnly=\"false\" userContent=\"true\" renameable=\"false\"><itemType>playlist</itemType><id>playlists</id><title>Playlists</title></mediaCollection><mediaCollection><itemType>container</itemType><id>genres</id><title>Genres</title></mediaCollection><mediaCollection><itemType>albumList</itemType><id>randomAlbums</id><title>Random</title></mediaCollection><mediaCollection><itemType>albumList</itemType><id>starredAlbums</id><title>Starred</title></mediaCollection><mediaCollection><itemType>albumList</itemType><id>recentlyAdded</id><title>Recently added</title></mediaCollection><mediaCollection><itemType>albumList</itemType><id>recentlyPlayed</id><title>Recently played</title></mediaCollection><mediaCollection><itemType>albumList</itemType><id>mostPlayed</id><title>Most played</title></mediaCollection></getMetadataResult></getMetadataResponse></soap:Body></soap:Envelope>"
  },
  "level": "debug",
  "service": "bonob",
  "timestamp": "2021-08-14 12:36:38"
}
```

This "looks" like the place where you would expect, in my case, a NL translation in the nodes.

Can't find any other reference to a language switch/selection in the logs.

**Comment by simojenki** [comment]
So there are a few things to check;

If you hit bonob at http://bonob/ you should see the index page in dutch, rather than english.  That works using the 'accept-language' header that the browser sends.  It should have nl-NL in it?

The 'Expected Config' should have a strings.version of >18.

The 'Existing service configuration' should have a matching strings.version, if not try register to update your sonos devices with the new config.

bonob itself does not caching, the sonos app does however cache.  So after registration above exit the app entirely, and re-open it.  It should then hit the http://bonob/sonos/strings.xml endpoint to get the latest translations.

I'll add some debug logging into the system so we can see what language headers your browser and sonos app are sending.  Perhaps they don't match quite what was expected from reading the docs.

**Comment by jicho** [comment]
> If you hit bonob at http://bonob/ you should see the index page in dutch, rather than english. That works using the 'accept-language' header that the browser sends. It should have nl-NL in it?

In Safari I see the EN version of the page, but in Chrome I see the NL version.
The ```Register``` button is not translated yet, that should be changed into ```Registreren``` or in this case ```Registreer``` is even better.

Back to the browsers. You're right about nl-NL, but it looks like there is a small difference. It there any chance you're comparing is case-sensitive?

Header in Chrome according to the browser debug console:
```Accept-Language: nl-NL,nl;q=0.9,en-US;q=0.8,en;q=0.7,de;q=0.6```

Header in Safari:
```Accept-Language: nl-nl```

Hmmm... and Firefox is doing it completely different:
```Accept-Language | nl,en-US;q=0.7,en;q=0.3```

So basically only in Chrome the translation works as expected.

> bonob itself does not caching, the sonos app does however cache. So after registration above exit the app entirely, and re-open it. It should then hit the http://bonob/sonos/strings.xml endpoint to get the latest translations.

Looking at ```http://bonob/sonos/strings.xml``` the response (in the browser) contains multiple languages where nl-NL is one of them, so that looks good.
 
> I'll add some debug logging into the system so we can see what language headers your browser and sonos app are sending. Perhaps they don't match quite what was expected from reading the docs.

Great, let me know when I update and check the debug info for you.

**Comment by simojenki** [comment]
I've added some additional debug so we can see what headers bonob is getting from both browsers and sonos app. I've also made the lang match case insensitive and support just the lang, without the region. ie. 'nl' == 'nl-NL' & 'en' == 'en-US' etc

**Comment by jicho** [comment]
The bonob page in my browser (Safari) is now in Dutch, that problem is solved :)

Looking at the overview (Artists, Albums, Starred, etc), that's still in English.
This is what I see in the logs:
```
{
   "message":"Fetching metadata type=root, typeId=undefined, acceptLanguage=nl-NL, en-US;q=0.9",
   "level":"debug",
   "service":"bonob",
   "timestamp":"2021-08-16 14:31:43"
}
```

Oh... And the "register" button still needs to receive the translation logic. Here is it still in English.

**Comment by simojenki** [comment]
Interesting that you do not see Dutch for the main menu as nl-NL is the expected lang.  I'll take a look tomorrow.

As for the 'regsiter' button being in English, that is because we have register="Register" in the nl-NL translations.   I'll update that to 'Registreren'

**Comment by simojenki** [comment]
ah, found a minor bug when determine the lang to use in main menu.  Test-case improved, bug fixed.

**Comment by jicho** [comment]
And... problem solved in the GUI!
I see Dutch text, great :)
